champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/ranger/axe/Agility.java

```diff
@@ -100,51 +100,65 @@ public double getCooldown(int level) {
 
     @EventHandler
     public void endOnInteract(PlayerInteractEvent event) {
-        if (event.getHand() != EquipmentSlot.HAND) return;
-        if (!event.getAction().isRightClick()) return;
-
         Player player = event.getPlayer();
-        if (active.contains(player.getUniqueId())) {
+        if (active.containsKey(player.getUniqueId())) {
             active.remove(player.getUniqueId());
-            player.removePotionEffect(PotionEffectType.SPEED);
+            deactivate(player);
         }
     }
 
-
     @EventHandler
     public void onDamage(CustomDamageEvent event) {
         if (!(event.getDamagee() instanceof Player damagee)) return;
-        if (!(event.getDamager() instanceof Player damager)) return;
-
-        if (active.contains(damagee.getUniqueId())) {
+        if (active.containsKey(damagee.getUniqueId())) {
             int level = getLevel(damagee);
             event.setDamage(event.getDamage() * (1 - getDamageReduction(level)));
             event.setKnockback(false);
-            UtilMessage.message(damager, getClassType().getName(), damagee.getName() + " is using " + getName());
+            if (event.getDamager() instanceof Player damager) {
+                UtilMessage.message(damager, getClassType().getName(), damagee.getName() + " is using " + getName());
+            }
             damagee.getWorld().playSound(damagee.getLocation(), Sound.ENTITY_BLAZE_AMBIENT, 0.5F, 2.0F);
         }
-
+        if (!(event.getDamager() instanceof Player damager)) return;
         if (event.getCause() == EntityDamageEvent.DamageCause.ENTITY_ATTACK) {
             active.remove(damager.getUniqueId());
+            deactivate(damager);
         }
-
     }
 
-    @UpdateEvent(delay = 500)
+    @UpdateEvent(delay = 250)
     public void onUpdate() {
-        active.removeIf(uuid -> Bukkit.getPlayer(uuid) == null);
+       Iterator<Map.Entry<UUID, Long>> iterator = active.entrySet().iterator();
+       while (iterator.hasNext()) {
+           Map.Entry<UUID, Long> entry = iterator.next();
+           Player player = Bukkit.getPlayer(entry.getKey());
+           if (player == null) {
+              iterator.remove();
+              continue;
+           } else {
+               if (entry.getValue() - System.currentTimeMillis() <= 0) {
+                   iterator.remove();
+                   deactivate(player);
```

Comment:
incredibly minor, but could you move iterator.remove() to below deactivate(player);